### ISSUE TITLE
TypeError: Cannot read property 'process' of undefined

### ISSUE THREAD

**Issue by fnando** [ai/autoprefixer-rails#47]
I'm trying to use autoprefixer-rails, but it's raising an exception. Here's the full backtrace:

```
TypeError: Cannot read property 'process' of undefined
execjs (2.2.2) lib/execjs/external_runtime.rb:64:in `extract_result'
execjs (2.2.2) lib/execjs/external_runtime.rb:29:in `exec'
execjs (2.2.2) lib/execjs/external_runtime.rb:18:in `eval'
execjs (2.2.2) lib/execjs/external_runtime.rb:36:in `call'
autoprefixer-rails (5.0.0) lib/autoprefixer-rails/processor.rb:21:in `process'
autoprefixer-rails (5.0.0) lib/autoprefixer-rails/sprockets.rb:15:in `process'
autoprefixer-rails (5.0.0) lib/autoprefixer-rails/sprockets.rb:21:in `block in install'
sprockets (2.12.3) lib/sprockets/processor.rb:29:in `call'
sprockets (2.12.3) lib/sprockets/processor.rb:29:in `evaluate'
tilt (1.4.1) lib/tilt/template.rb:103:in `render'
sprockets (2.12.3) lib/sprockets/context.rb:197:in `block in evaluate'
sprockets (2.12.3) lib/sprockets/context.rb:194:in `each'
sprockets (2.12.3) lib/sprockets/context.rb:194:in `evaluate'
sprockets (2.12.3) lib/sprockets/processed_asset.rb:12:in `initialize'
sprockets (2.12.3) lib/sprockets/base.rb:374:in `new'
sprockets (2.12.3) lib/sprockets/base.rb:374:in `block in build_asset'
sprockets (2.12.3) lib/sprockets/base.rb:395:in `circular_call_protection'
sprockets (2.12.3) lib/sprockets/base.rb:373:in `build_asset'
sprockets (2.12.3) lib/sprockets/index.rb:94:in `block in build_asset'
sprockets (2.12.3) lib/sprockets/caching.rb:58:in `cache_asset'
sprockets (2.12.3) lib/sprockets/index.rb:93:in `build_asset'
sprockets (2.12.3) lib/sprockets/base.rb:287:in `find_asset'
sprockets (2.12.3) lib/sprockets/index.rb:61:in `find_asset'
sprockets (2.12.3) lib/sprockets/bundled_asset.rb:16:in `initialize'
sprockets (2.12.3) lib/sprockets/base.rb:377:in `new'
sprockets (2.12.3) lib/sprockets/base.rb:377:in `build_asset'
sprockets (2.12.3) lib/sprockets/index.rb:94:in `block in build_asset'
sprockets (2.12.3) lib/sprockets/caching.rb:58:in `cache_asset'
sprockets (2.12.3) lib/sprockets/index.rb:93:in `build_asset'
sprockets (2.12.3) lib/sprockets/base.rb:287:in `find_asset'
sprockets (2.12.3) lib/sprockets/index.rb:61:in `find_asset'
sprockets (2.12.3) lib/sprockets/environment.rb:75:in `find_asset'
sprockets (2.12.3) lib/sprockets/base.rb:295:in `[]'
sprockets-rails (2.2.2) lib/sprockets/rails/helper.rb:230:in `lookup_asset_for_path'
sprockets-rails (2.2.2) lib/sprockets/rails/helper.rb:190:in `check_errors_for'
sprockets-rails (2.2.2) lib/sprockets/rails/helper.rb:159:in `block in stylesheet_link_tag'
sprockets-rails (2.2.2) lib/sprockets/rails/helper.rb:158:in `map'
sprockets-rails (2.2.2) lib/sprockets/rails/helper.rb:158:in `stylesheet_link_tag'
app/views/layouts/application.html.erb:5:in `_app_views_layouts_application_html_erb___4471141664269420399_70229777939800'
actionview (4.2.0) lib/action_view/template.rb:145:in `block in render'
activesupport (4.2.0) lib/active_support/notifications.rb:166:in `instrument'
actionview (4.2.0) lib/action_view/template.rb:333:in `instrument'
actionview (4.2.0) lib/action_view/template.rb:143:in `render'
actionview (4.2.0) lib/action_view/renderer/template_renderer.rb:66:in `render_with_layout'
actionview (4.2.0) lib/action_view/renderer/template_renderer.rb:52:in `render_template'
actionview (4.2.0) lib/action_view/renderer/template_renderer.rb:14:in `render'
actionview (4.2.0) lib/action_view/renderer/renderer.rb:42:in `render_template'
actionview (4.2.0) lib/action_view/renderer/renderer.rb:23:in `render'
actionview (4.2.0) lib/action_view/rendering.rb:100:in `_render_template'
actionpack (4.2.0) lib/action_controller/metal/streaming.rb:217:in `_render_template'
actionview (4.2.0) lib/action_view/rendering.rb:83:in `render_to_body'
actionpack (4.2.0) lib/action_controller/metal/rendering.rb:32:in `render_to_body'
actionpack (4.2.0) lib/action_controller/metal/renderers.rb:37:in `render_to_body'
actionpack (4.2.0) lib/abstract_controller/rendering.rb:25:in `render'
actionpack (4.2.0) lib/action_controller/metal/rendering.rb:16:in `render'
actionpack (4.2.0) lib/action_controller/metal/instrumentation.rb:41:in `block (2 levels) in render'
activesupport (4.2.0) lib/active_support/core_ext/benchmark.rb:12:in `block in ms'
/Users/fvieira/local/ruby/ruby-2.2.0/lib/ruby/2.2.0/benchmark.rb:303:in `realtime'
activesupport (4.2.0) lib/active_support/core_ext/benchmark.rb:12:in `ms'
actionpack (4.2.0) lib/action_controller/metal/instrumentation.rb:41:in `block in render'
actionpack (4.2.0) lib/action_controller/metal/instrumentation.rb:84:in `cleanup_view_runtime'
actionpack (4.2.0) lib/action_controller/metal/instrumentation.rb:40:in `render'
actionpack (4.2.0) lib/action_controller/metal/implicit_render.rb:10:in `default_render'
actionpack (4.2.0) lib/action_controller/metal/implicit_render.rb:5:in `send_action'
actionpack (4.2.0) lib/abstract_controller/base.rb:198:in `process_action'
actionpack (4.2.0) lib/action_controller/metal/rendering.rb:10:in `process_action'
actionpack (4.2.0) lib/abstract_controller/callbacks.rb:20:in `block in process_action'
activesupport (4.2.0) lib/active_support/callbacks.rb:117:in `call'
activesupport (4.2.0) lib/active_support/callbacks.rb:117:in `call'
activesupport (4.2.0) lib/active_support/callbacks.rb:234:in `block in halting'
activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `call'
activesupport (4.2.0) lib/active_support/callbacks.rb:169:in `block in halting'
activesupport (4.2.0) lib/active_support/callbacks.rb:92:in `call'
activesupport (4.2.0) lib/active_support/callbacks.rb:92:in `_run_callbacks'
activesupport (4.2.0) lib/active_support/callbacks.rb:734:in `_run_process_action_callbacks'
activesupport (4.2.0) lib/active_support/callbacks.rb:81:in `run_callbacks'
actionpack (4.2.0) lib/abstract_controller/callbacks.rb:19:in `process_action'
actionpack (4.2.0) lib/action_controller/metal/rescue.rb:29:in `process_action'
actionpack (4.2.0) lib/action_controller/metal/instrumentation.rb:31:in `block in process_action'
activesupport (4.2.0) lib/active_support/notifications.rb:164:in `block in instrument'
activesupport (4.2.0) lib/active_support/notifications/instrumenter.rb:20:in `instrument'
activesupport (4.2.0) lib/active_support/notifications.rb:164:in `instrument'
actionpack (4.2.0) lib/action_controller/metal/instrumentation.rb:30:in `process_action'
actionpack (4.2.0) lib/action_controller/metal/params_wrapper.rb:250:in `process_action'
actionpack (4.2.0) lib/abstract_controller/base.rb:137:in `process'
actionview (4.2.0) lib/action_view/rendering.rb:30:in `process'
actionpack (4.2.0) lib/action_controller/metal.rb:195:in `dispatch'
actionpack (4.2.0) lib/action_controller/metal/rack_delegation.rb:13:in `dispatch'
actionpack (4.2.0) lib/action_controller/metal.rb:236:in `block in action'
actionpack (4.2.0) lib/action_dispatch/routing/route_set.rb:73:in `call'
actionpack (4.2.0) lib/action_dispatch/routing/route_set.rb:73:in `dispatch'
actionpack (4.2.0) lib/action_dispatch/routing/route_set.rb:42:in `serve'
actionpack (4.2.0) lib/action_dispatch/journey/router.rb:43:in `block in serve'
actionpack (4.2.0) lib/action_dispatch/journey/router.rb:30:in `each'
actionpack (4.2.0) lib/action_dispatch/journey/router.rb:30:in `serve'
actionpack (4.2.0) lib/action_dispatch/routing/route_set.rb:802:in `call'
rack (1.6.0) lib/rack/etag.rb:24:in `call'
rack (1.6.0) lib/rack/conditionalget.rb:25:in `call'
rack (1.6.0) lib/rack/head.rb:13:in `call'
actionpack (4.2.0) lib/action_dispatch/middleware/params_parser.rb:27:in `call'
actionpack (4.2.0) lib/action_dispatch/middleware/flash.rb:260:in `call'
rack (1.6.0) lib/rack/session/abstract/id.rb:225:in `context'
rack (1.6.0) lib/rack/session/abstract/id.rb:220:in `call'
actionpack (4.2.0) lib/action_dispatch/middleware/cookies.rb:560:in `call'
actionpack (4.2.0) lib/action_dispatch/middleware/callbacks.rb:29:in `block in call'
activesupport (4.2.0) lib/active_support/callbacks.rb:88:in `call'
activesupport (4.2.0) lib/active_support/callbacks.rb:88:in `_run_callbacks'
activesupport (4.2.0) lib/active_support/callbacks.rb:734:in `_run_call_callbacks'
activesupport (4.2.0) lib/active_support/callbacks.rb:81:in `run_callbacks'
actionpack (4.2.0) lib/action_dispatch/middleware/callbacks.rb:27:in `call'
actionpack (4.2.0) lib/action_dispatch/middleware/reloader.rb:73:in `call'
actionpack (4.2.0) lib/action_dispatch/middleware/remote_ip.rb:78:in `call'
actionpack (4.2.0) lib/action_dispatch/middleware/debug_exceptions.rb:17:in `call'
actionpack (4.2.0) lib/action_dispatch/middleware/show_exceptions.rb:30:in `call'
railties (4.2.0) lib/rails/rack/logger.rb:38:in `call_app'
railties (4.2.0) lib/rails/rack/logger.rb:20:in `block in call'
activesupport (4.2.0) lib/active_support/tagged_logging.rb:68:in `block in tagged'
activesupport (4.2.0) lib/active_support/tagged_logging.rb:26:in `tagged'
activesupport (4.2.0) lib/active_support/tagged_logging.rb:68:in `tagged'
railties (4.2.0) lib/rails/rack/logger.rb:20:in `call'
actionpack (4.2.0) lib/action_dispatch/middleware/request_id.rb:21:in `call'
rack (1.6.0) lib/rack/methodoverride.rb:22:in `call'
rack (1.6.0) lib/rack/runtime.rb:18:in `call'
activesupport (4.2.0) lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
rack (1.6.0) lib/rack/lock.rb:17:in `call'
actionpack (4.2.0) lib/action_dispatch/middleware/static.rb:113:in `call'
rack (1.6.0) lib/rack/sendfile.rb:113:in `call'
railties (4.2.0) lib/rails/engine.rb:518:in `call'
railties (4.2.0) lib/rails/application.rb:164:in `call'
rack (1.6.0) lib/rack/lock.rb:17:in `call'
rack (1.6.0) lib/rack/content_length.rb:15:in `call'
rack (1.6.0) lib/rack/handler/webrick.rb:89:in `service'
/Users/fvieira/local/ruby/ruby-2.2.0/lib/ruby/2.2.0/webrick/httpserver.rb:138:in `service'
/Users/fvieira/local/ruby/ruby-2.2.0/lib/ruby/2.2.0/webrick/httpserver.rb:94:in `run'
/Users/fvieira/local/ruby/ruby-2.2.0/lib/ruby/2.2.0/webrick/server.rb:294:in `block in start_thread'
```

Info about my environment:

```
$ uname -a
Darwin fnando 14.0.0 Darwin Kernel Version 14.0.0: Fri Sep 19 00:26:44 PDT 2014; root:xnu-2782.1.97~2/RELEASE_X86_64 x86_6

$ ruby -v
ruby 2.2.0p0 (2014-12-25 revision 49005) [x86_64-darwin14]

$ node -v
v0.10.35

$ bundle show
Gems included by the bundle:
  * actionmailer (4.2.0)
  * actionpack (4.2.0)
  * actionview (4.2.0)
  * activejob (4.2.0)
  * activemodel (4.2.0)
  * activerecord (4.2.0)
  * activesupport (4.2.0)
  * arel (6.0.0)
  * autoprefixer-rails (5.0.0)
  * builder (3.2.2)
  * bundler (1.7.11)
  * erubis (2.7.0)
  * execjs (2.2.2)
  * globalid (0.3.0)
  * hike (1.2.3)
  * i18n (0.7.0)
  * json (1.8.2)
  * loofah (2.0.1)
  * mail (2.6.3)
  * mime-types (2.4.3)
  * mini_portile (0.6.2)
  * minitest (5.5.1)
  * multi_json (1.10.1)
  * nokogiri (1.6.5)
  * rack (1.6.0)
  * rack-test (0.6.3)
  * rails (4.2.0)
  * rails-deprecated_sanitizer (1.0.3)
  * rails-dom-testing (1.0.5)
  * rails-html-sanitizer (1.0.1)
  * railties (4.2.0)
  * rake (10.4.2)
  * sass (3.4.9)
  * sass-rails (5.0.1)
  * sprockets (2.12.3)
  * sprockets-rails (2.2.2)
  * thor (0.19.1)
  * thread_safe (0.3.4)
  * tilt (1.4.1)
  * tzinfo (1.2.2)
  * uglifier (2.7.0)
```

The application I tried can be downloaded here: https://dl.dropbox.com/s/81un2gpgb5e6o1p/autoprefixer-bug-tar.gz

**Comment by fnando** [comment]
FYI, v4.0.2.2 doesn't raise this exception.

**Comment by twalpole** [comment]
FYI - this only happens for me when ExecJS is using nodejs as its runtime.  The error does not occur when using therubyracer as the runtime

**Comment by ai** [comment]
Hm. Very bad error, something wrong in my ExecJS JS around Autoprefixer.

Error should be somewhere here: https://github.com/ai/autoprefixer-rails/blob/master/lib/autoprefixer-rails/processor.rb#L110

Does anyone have a ideas?

**Comment by mattbrictson** [comment]
@ai I think the problem is that [runtime.exec on line 20](https://github.com/ai/autoprefixer-rails/blob/master/lib/autoprefixer-rails/processor.rb#L20) has no effect on the state of the runtime. Therefore [runtime.call on line 21](https://github.com/ai/autoprefixer-rails/blob/master/lib/autoprefixer-rails/processor.rb#L21) still sees a value of `undefined` for the `processor` variable. If you look at the [ExecJS code](https://github.com/sstephenson/execjs/blob/master/lib/execjs/external_runtime.rb#L25) you'll see that each call to `eval` or `call` generates a new temp file that is evaluated by e.g. Node each time.

So I think the code needs to be rewritten such that `def process(css, opts={})` makes a single call to the JS runtime, rather than `eval` followed by `call`.

**Comment by ai** [comment]
@mattbrictson maybe you can help and send PR? Because I will have time only on this or next weekend.

**Comment by clemensg** [comment]
Please also add a regression/integration-test with both therubyracer and node as runtime to prevent similar bugs in the future.

**Comment by ai** [comment]
@clemensg can you send PR with `.travis.yml` for this tests?

**Comment by tagliala** [comment]
I agree with @clemensg , version 5.0.0 should be yanked

I'm sorry but I don't have the knowledge to help here :disappointed:

**Comment by ai** [comment]
I try to fix tomorrow. If I will have not solution until evening, I will
yank the version.
15 янв. 2015 г. 18:55 пользователь "Geremia Taglialatela" <
notifications@github.com> написал:

**Comment by twalpole** [comment]
PR #48 should fix this

**Comment by ai** [comment]
@twalpole BTW, what is yoour Twitter account to mention you in new release tweet?

**Comment by twalpole** [comment]
@ai don't have one - not a big social media person

**Comment by tagliala** [comment]
thanks @ai @twalpole @mattbrictson  :+1:

**Comment by androidand** [comment]
I am learning Ruby on rails and was experimenting by trying to add sass and prefixer to my app, but got the error stated above "TypeError: Cannot read property 'process' of undefined" and had no clue what was wrong until I saw the comment by fnando "FYI, v4.0.2.2 doesn't raise this exception." Added gem 'autoprefixer-rails', '~> 4.0.2.2' to Gemfile and ran bundler, now my app works again. Thank you and hopefully someone will fix version 5.

**Comment by ai** [comment]
@androidand 5.0 is already fixed in 5.0.0.1 release. Run `bundle update autoprefixer-rails`

**Comment by androidand** [comment]
Cool, thank you, I just did and can confirm it works. Good job.
